4th task of testing

IvayloTsvetkov opened the pull request:
Tests for EndpointFilter, implementation of StartsWithKeyword and tests for it.

mgenov commented on testing_4/test/com/company/EndpointFilterTest.java:
```diff
@@ -0,0 +1,39 @@
+  public JUnitRuleMockery context = new JUnitRuleMockery();
+
+  @Test
+  public void shouldFilterReturnsTrue() {
```
filterMatches filterNotMatches sounds better then explicitly describing what is going to be returned.

mgenov commented on testing_4/test/com/company/StartsWithKeywordTest.java:
```diff
@@ -0,0 +1,26 @@
+import static org.junit.Assert.*;
+
+public class StartsWithKeywordTest {
+  private StartsWithKeyword startsWithKeyword;
```
The EndpointFiler should work with many endpoints, so the scenarios should be different:

1. zero endpoints
2. two endpoints and first matches
3. two endpoints and second matches
4. multiple endpoints and none matches
5. multiple endpoints and one matches

I don't explicitly mention scenarios with 1 endpoint as I assume that they overlap with case for many endpoints.

IvayloTsvetkov commented:
Fixed some method names and added more tests in StartsWithKeywordTest to cover all cases. And I'd like to ask if i should write these tests for EndpointFilterTest too or is it going to make the code redundant ?

mgenov commented on testing_4/test/com/company/StartsWithKeywordTest.java:
```diff
@@ -0,0 +1,70 @@
+  @Test
+  public void endpointMatches() {
+    StartsWithKeyword startsWithKeyword = new StartsWithKeyword("add");
+    EndpointFilter endpointFilter = new EndpointFilter(startsWithKeyword);
```
Testing of EndpointFilter doesn't have to use StartsWithKeyword as it tests them in integrated manner. 

Integrated tests are scam ! 

http://blog.thecodewhisperer.com/permalink/integrated-tests-are-a-scam

For testing of EndpointFilter you have to use Mock or Fake object instead.

IvayloTsvetkov commented:
Integrated tests are now removed.

mgenov commented on testing_4/test/com/company/EndpointFilterTest.java:
```diff
@@ -0,0 +1,119 @@
+
+import static org.junit.Assert.*;
+
+public class EndpointFilterTest {
```
Add another test class which uses Fake/Dummy object for isolation and write some lines about differences. Consider which option is better for testing of the EndpointFilter.

IvayloTsvetkov replied:
I just saw how write tests using fake/dummy, but i'm not sure if i have to use StartsWithKeyword for the expected answer or maybe use mock along with dummy ? I'm a little bit confused in this regard.

mgenov replied:
Dummy/Fake object need to be used. StartsWithKeyword is concrete implementation which should not be used for testing of the EndpointFilter.

IvayloTsvetkov commented:
I pushed the solution i came up with, but i'm not sure if it's correct.

mgenov commented:
I'm not able to see the new test that uses Dummy/Fake objects. Only the implementation with jmock is available now.

IvayloTsvetkov commented:
Sorry, it's now up under EndpointFilterTest2 name.

mgenov commented on testing_4/test/com/company/EndpointFilterTest2.java:
```diff
@@ -0,0 +1,52 @@
+import static org.junit.Assert.*;
+
+public class EndpointFilterTest2 {
+  class EndpointDummy implements Endpoint {
```
Improve Name: class DummyEndpoint implements Endpoint 

You can pass to the constructor endpoint the return value

new DummyEndpoint(false)
new DummyEndpoint(true) 

For better reading you can create methods which to return proper instances: 

```java
DummyEndpoint notMatchingEndpoint = newNotMatchingEndpoint()
DummyEndpoint matchingEndpoint = newMatchingEndpoint()
EndpointFilter endpointFilter = new EndpointFilter(notMatchingEndpoint, matchingEndpoint) 
asserTrue(endpointFilter.shouldFilter("address")) 
assertThat(notMatchingEndpoint.usedCount, is(equalTo(1));
```

I've added usedCount as you are not verified whether all of the members are used. This could be verified and with scenarios but this makes test more explicit.

IvayloTsvetkov replied:
usedCount seems good if i have only one notMatchingEndpoint, but if i have two, which i have then i have to either add another assertThat or implement the counter in shouldFilter(). So how should i proceed with that ?

mgenov replied:
You need to figure out the way for testing it out and to ensure that tests is verifying correctly the behaviour of the implementation.

IvayloTsvetkov replied:
If i have 3 endpoints and only the last one matches, is it enough if i verify only 1 of the non matching endpoints was used ?

mgenov replied:
You may, but you are not sure that the other one is called only once. 

```java
if (firstMatcher.matches(value)) {
  return true;
}
secondMatcher.matches(value);
return secondMatcher.matches(value);
```

What will your test show about this implementation ?

IvayloTsvetkov replied:
I can come with plenty of solutions, but i don't know which one is right. Is it all right if i use the same nonMatchingEndpoint twice and then assert that count is equal 2 ?
```
DummyEndpoint notMatchingEndpoint = newNotMatchingEndpoint();
    DummyEndpoint matchingEndpoint = newMatchingEndpoint();
    EndpointFilter endpointFilter = new EndpointFilter(notMatchingEndpoint,
        , notMatchingEndpoint, matchingEndpoint);
    assertTrue(endpointFilter.shouldFilter("address"));
    assertThat(notMatchingEndpoint.getCount(), equalTo(2));
```

mgenov replied:
You don't have to assert count is 2 as you need to ensure that some are called once and some may not be called as some before them was already matching.

IvayloTsvetkov commented:
I added verification for each endpoint, which is probably not the way to go.

mgenov commented:
You are verifying single thing many times. You can make one or 2 tests which are just verifying that.

IvayloTsvetkov commented:
How can i do this ? Is it with some hamcrest matcher that I don't know about ?

IvayloTsvetkov commented:
Pushed
